janusgraph-server/src/main/java/org/janusgraph/graphdb/server/JanusGraphServer.java

```diff
@@ -17,52 +17,86 @@
 import com.jcabi.manifests.Manifests;
 import org.apache.tinkerpop.gremlin.server.GremlinServer;
 import org.apache.tinkerpop.gremlin.server.Settings;
-import org.janusgraph.graphdb.server.utils.JanusGraphSettingsUtils;
+import org.janusgraph.graphdb.server.utils.GremlinSettingsUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.concurrent.CompletableFuture;
+
 public class JanusGraphServer {
     private static final Logger logger = LoggerFactory.getLogger(JanusGraphServer.class);
+    private GremlinServer gremlinServer = null;
+    private Settings gremlinSettings = null;
+    private final String confPath;
+    private CompletableFuture<Void> serverStarted = null;
+    private CompletableFuture<Void> serverStopped = null;
+
+    public JanusGraphServer(final String file) {
+        confPath = file;
+    }
 
     public static void main(final String[] args) throws Exception {
         printHeader();
-        final String file;
-        if (args.length > 0)
-            file = args[0];
-        else
-            file = "conf/janusgraph-server.yaml";
-
-        final Settings settings;
-        try {
-            settings = JanusGraphSettingsUtils.configureDefaultSerializersIfNotSet(Settings.read(file));
-        } catch (Exception ex) {
-            logger.error("Configuration file at {} could not be found or parsed properly. [{}]", file, ex.getMessage());
-            return;
-        }
-
-        logger.info("Configuring JanusGraph Server from {}", file);
-        final GremlinServer server = new GremlinServer(settings);
-        server.start().exceptionally(t -> {
+        final String file = (args.length > 0) ? args[0] : "conf/janusgraph-server.yaml";
+        JanusGraphServer janusGraphServer = new JanusGraphServer(file);
+        janusGraphServer.start().exceptionally(t -> {
             logger.error("JanusGraph Server was unable to start and will now begin shutdown: {}", t.getMessage());
-            server.stop().join();
+            janusGraphServer.stop().join();
             return null;
         }).join();
     }
 
+    public GremlinServer getGremlinServer() {
+        return gremlinServer;
+    }
+
+    public Settings getGremlinSettings() {
+        return gremlinSettings;
+    }
+
+    public synchronized CompletableFuture<Void> start() {
+        if (serverStarted != null) {
+            return serverStarted;
+        }
+        serverStarted = new CompletableFuture<>();
+        try
+        {
+            logger.info("Configuring JanusGraph Server from {}", confPath);
+            gremlinSettings = GremlinSettingsUtils.configureDefaultSerializersIfNotSet(Settings.read(confPath));
+            gremlinServer = new GremlinServer(gremlinSettings);
+            serverStarted = CompletableFuture.allOf(gremlinServer.start());
```

Comment:
From my understanding it is equivalent to the following:
```suggestion
            serverStarted = gremlinServer.start();
```

Reply:
Isn't the return type different.

Reply:
Yeah, my bad